File: WordCounter.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;

class WordCounter
{
    static Stopwatch stopwatch = new Stopwatch();
    static void Main()
    {
        string filePath = GetValidFilePath();
        var wordCounts = new Dictionary<string, int>();

        stopwatch.Start();
        using (var reader = new StreamReader(filePath))
        {
            while (!reader.EndOfStream)
            {
                string line = reader.ReadLine();
                CountWordsInLine(line, wordCounts);
            }
        }
        stopwatch.Stop();

        PrintWordCounts(wordCounts);
        MeasureTimeAndMemory();
    }
    static void CountWordsInLine(string line, Dictionary<string, int> wordCounts)
    {
        char[] delimiters = { ' ', '\n', '\t'};
        string[] words = line.Split(delimiters)
                            .Select(word => Regex.Replace(word, @"\W+", "").ToLower())
                            .Where(word => !string.IsNullOrWhiteSpace(word))
                            .ToArray();

        foreach (string word in words)
        {
            if (wordCounts.ContainsKey(word))
            {
                wordCounts[word]++;
            }
            else
            {
                wordCounts[word] = 1;
            }
        }
    }

    static void PrintWordCounts(Dictionary<string, int> wordCounts)
    {
        foreach (var wordCount in wordCounts.OrderBy(w => w.Key))
        {
            Console.WriteLine("{0}: {1}", wordCount.Key, wordCount.Value);
        }
    }

    static string GetValidFilePath()
    {
        string filePath = "";
        bool isValidFilePath = false;

        while (!isValidFilePath)
        {
            Console.Write("Enter the path to the input file: ");
            filePath = Console.ReadLine();

            if (File.Exists(filePath))
            {
                isValidFilePath = true;
            }
            else
            {
                Console.WriteLine("Invalid file path. Please enter a valid file path.");
            }
        }

        return filePath;
    }

     static void MeasureTimeAndMemory()
    {
        long memoryUsage = Process.GetCurrentProcess().WorkingSet64;
        Console.WriteLine("Memory usage: {0:N0} MB", memoryUsage / 1000000);

        TimeSpan elapsedTime = stopwatch.Elapsed;
        Console.WriteLine("Execution time: {0:F0} ms", elapsedTime.TotalMilliseconds);
    }

}

File: WordCounter.scala
import scala.collection.immutable.SortedMap
import scala.io.Source
import scala.io.StdIn.readLine
import scala.util.matching.Regex
import scala.concurrent.duration._

object WordCounter {
  def main(args: Array[String]): Unit = {
    val filePath = readLine("Enter file path: ")
    
    val (wordCounts, time, memory) = try {
      measureTimeAndMemory(countWords(Source.fromFile(filePath).getLines()))
    } finally {
      println("File closed.")
    }
    
    printWordCounts(wordCounts)
    
    println(s"Execution time: ${time.toMillis} ms")
    println(s"Memory usage: ${memory / 1000000} MB")
  }

  def countWords(lines: Iterator[String]): SortedMap[String, Int] = {
    val delimiters = Array(' ','\n', '\t')
    val regex = new Regex("\\W+")

    lines.flatMap(line => line.split(delimiters))
         .map(word => regex.replaceAllIn(word.toLowerCase(), ""))
         .filter(word => word.nonEmpty)
         .foldLeft(SortedMap.empty[String, Int])((counts, word) =>
        counts.updated(word, counts.getOrElse(word, 0) + 1))
  }

  def printWordCounts(wordCounts: SortedMap[String, Int]): Unit = {
    for ((word, count) <- wordCounts) {
      println(s"$word: $count")
    }
  }

  def measureTimeAndMemory[T](block: => T): (T, FiniteDuration, Long) = {
    val start = System.nanoTime()
    val result = block
    val time = (System.nanoTime() - start).nanos
    val memory = Runtime.getRuntime.totalMemory() - Runtime.getRuntime.freeMemory()
    (result, time, memory)
  }
}
